MM-53374: Can make API calls when the custom terms of service are not accepted

catalintomai opened the pull request:
#### Summary
Can make API calls when the custom terms of service are not accepted.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53374

#### Release Note

```release-note
NONE
```

catalintomai committed:
initial checkin

catalintomai committed:
update test

catalintomai committed:
update test(1)

catalintomai committed:
update test(2)

catalintomai committed:
fix logic

catalintomai committed:
update logic

catalintomai committed:
update logic(1)

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
add caching

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
small improvs

catalintomai committed:
further improvs

catalintomai committed:
further improvs

catalintomai committed:
fix test

agarciamontoro commented on server/channels/api4/user_test.go:
```diff
@@ -9446,3 +9446,78 @@ func TestSearchUsersWithMfaEnforced(t *testing.T) {
+		require.NoError(t, err)
+	})
+
+	t.Run("API calls allowed after accepting ToS", func(t *testing.T) {
+		// Now API calls should work
+		_, _, err := th.Client.GetUser(context.Background(), th.BasicUser.Id, "")
+		require.NoError(t, err)
+
+		_, _, err = th.Client.GetMe(context.Background(), "")
+		require.NoError(t, err)
+	})
```
Nit: this test depends on the previous ones. Would it be possible to decouple them?

agarciamontoro commented on server/channels/store/localcachelayer/user_terms_of_service_layer_test.go:
```diff
@@ -0,0 +1,85 @@
+	fakeUserTermsOfService := model.UserTermsOfService{UserId: "user123", TermsOfServiceId: "123", CreateAt: 11111}
+	logger := mlog.CreateConsoleTestLogger(t)
+
+	t.Run("first call not cached, second cached and returning same data", func(t *testing.T) {
+		mockStore := getMockStore(t)
+		mockCacheProvider := getMockCacheProvider()
+		cachedStore, err := NewLocalCacheLayer(mockStore, nil, nil, mockCacheProvider, logger)
+		require.NoError(t, err)
+
+		userTermsOfService, err := cachedStore.UserTermsOfService().GetByUser("user123")
+		require.NoError(t, err)
+		assert.Equal(t, userTermsOfService, &fakeUserTermsOfService)
+		mockStore.UserTermsOfService().(*mocks.UserTermsOfServiceStore).AssertNumberOfCalls(t, "GetByUser", 1)
+
+		userTermsOfService, err = cachedStore.UserTermsOfService().GetByUser("user123")
+		require.NoError(t, err)
+		assert.Equal(t, userTermsOfService, &fakeUserTermsOfService)
+		mockStore.UserTermsOfService().(*mocks.UserTermsOfServiceStore).AssertNumberOfCalls(t, "GetByUser", 1)
+	})
```
Is this test really checking that the second call returned cached data? Shouldn't we be asserting 0 on the second `AssertNumberOfCalls`?

catalintomai replied:
AssertNumberOfCalls(t, "GetByUser", 1) means "assert that the method has been called exactly 1 time total" - not "assert that it was called 1 time since the last check."

agarciamontoro replied:
Oooh, got it! Thanks for the explanation :)

agarciamontoro reviewed (changes requested):
Amazing work, fascinating that we were not checking this until now! I left some comments below.

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
address comments

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
Merge branch 'master' into SET-MM-53374

lieut-data commented on server/channels/web/context.go:
```diff
@@ -139,6 +143,119 @@ func (c *Context) SessionRequired() {
+	return slices.Contains(exemptPaths, cleanPath)
+}
+
+// matchesUserToSPattern checks if path matches the pattern /api/v4/users/{user_id}/terms_of_service
+// using the same validation pattern as Gorilla Mux: {user_id:[A-Za-z0-9]+}
```
Having to copy the Gorilla Mux validation pattern suggests to me this logic belongs in `api4/` not `web/`.

Can we extend the `api.APISessionRequired` with an option to flag those APIs that can always be used when TOS is enforced? See `RequireMfa` as a comparable.

catalintomai committed:
address comments

catalintomai committed:
address comments

catalintomai committed:
fix spacing

catalintomai committed:
fix spacing

catalintomai committed:
remove duplication

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai commented:
> > @lieut-data, should I re-assign the review of this SET ticket to somebody else? Thanks.
> 
> Yes, please assign to me @catalintomai. Thanks!

Think it was assigned to you initially = ).

catalintomai commented:
> Apologies, @catalintomai -- I thought you meant ownership of the ticket, as this has obviously gone well beyond your SET rotation. Happy to help drive it across the line :)

At the current pace, it is possible these will be closed at my next rotation = ).

sbishel reviewed (approved):
Approving, just a minor comment.

lindy65 reviewed (approved):
Thanks @catalintomai - this is fixed

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai commented:
@esarafianou , I assume your team would like to review this as well? Thanks.

catalintomai committed:
address comments

esarafianou commented:
@catalintomai, yes I'll review this week.

esarafianou commented on server/channels/app/user_terms_of_service.go:
```diff
@@ -46,6 +48,18 @@ func (a *App) SaveUserTermsOfService(userID, termsOfServiceId string, accepted b
+		// This ensures that any cached ToS acceptance in existing sessions is invalidated
+		c := request.EmptyContext(a.Log())
+		if appErr := a.RevokeAllSessions(c, userID); appErr != nil {
+			// Log the error but don't fail the ToS deletion - session revocation is a security enhancement
```
@catalintomai what do we mean by security enhancement? 

If there's no security issue with session revocation failing, then why perform a user disruptive operation such as revoking all their sessions?

If there is a way for a user to bypass ToS acceptance if the session revocation fails, then this is a security issue.

lieut-data replied:
@esarafianou, I've opted to remove the session level caching and only cache at the store level to simplify things. Let me know if you have any concerns!

esarafianou reviewed (changes requested):
Marking as requesting changes, mostly to address the two comments.

catalintomai committed:
Merge branch 'master' into SET-MM-53374

esarafianou commented on server/channels/web/handlers.go:
```diff
@@ -319,6 +322,12 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		c.SessionRequired()
 	}
 
+	if c.Err == nil && h.RequireSession && !h.SkipTermsOfService {
+		if appErr := c.TermsOfServiceRequired(r); appErr != nil {
+			c.Err = appErr
+		}
```
MFA should happen first. If MFA is required and hasn't been completed, we don't have an authenticated user to show and accept ToS

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
address comments

catalintomai committed:
fix check-style

catalintomai committed:
Merge branch 'master' into SET-MM-53374

catalintomai committed:
add defensive check

lieut-data committed:
Merge remote-tracking branch 'origin/master' into SET-MM-53374

# Conflicts:
#	server/channels/store/localcachelayer/layer.go
#	server/channels/web/handlers_test.go
#	server/public/model/cluster_message.go

esarafianou commented:
@lieut-data, @catalintomai is this work still relevant? There was an ongoing discussion around security that it seems I missed following up. Let me know if this PR is still relevant and we can resume discussions.

lieut-data committed:
Merge remote-tracking branch 'origin/master' into SET-MM-53374

lieut-data committed:
Simplify ToS enforcement to use store-level caching only

Remove session-level caching of ToS acceptance status in favor of
the UserTermsOfService store cache layer. This eliminates complexity
around session prop synchronization and cache invalidation.

Changes:
- Remove SessionPropTermsOfServiceId constant and session prop usage
- Remove session revocation on ToS rejection (store cache handles invalidation)
- Simplify TermsOfServiceRequired to always query store (which is cached)

esarafianou reviewed (approved):
Thanks Jesse. Just a question below.

lindy65 commented:
@catalintomai - I've removed myself from review for now. Please re-add me when you resume work on this.

lieut-data commented:
Closing out this PR. We're still tracking internally as https://mattermost.atlassian.net/browse/MM-53374 and will revisit, but not merging these changes at this time.
